File: RewardRandomizer.Tests/IPSTests.cs
using Microsoft.FSharp.Collections;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Diagnostics;

namespace RewardRandomizer.Tests
{
    [TestClass]
    public class IPSTests
    {
        [TestMethod]
        public void TestGeneration()
        {
            var random = new Random();
            var data1 = new byte[0x100];
            random.NextBytes(data1);
            var operations = new[]
            {
                new Randomizer.WriteOperation(offsets: SetModule.OfArray(new[] { 0x30, 0x50, 0x51 }), writeData: 7),
                new Randomizer.WriteOperation(offsets: SetModule.OfArray(new[] { 0x99 }), writeData: 100),
            };
            if (!File.Exists("Lunar IPS.exe"))
                Assert.Inconclusive();
            File.WriteAllBytes("original_file", data1);
            File.WriteAllBytes("patch_file", Randomizer.CreateIPS(operations));
            Process.Start("Lunar IPS.exe", "-ApplyIPS patch_file original_file").WaitForExit();
            var expected = File.ReadAllBytes("original_file");
            File.Delete("original_file");
            File.Delete("patch_file");
            var actual = Randomizer.ApplyOperations(data1, operations);
            Assert.AreEqual(
                ListModule.OfArray(expected),
                ListModule.OfArray(actual));
        }

        [TestMethod, ExpectedException(typeof(Exception))]
        public void TestProhibitAmbiguity()
        {
            var random = new Random();
            var data1 = new byte[0x500000];
            random.NextBytes(data1);
            var operations = new[]
            {
                new Randomizer.WriteOperation(offsets: SetModule.OfArray(new[] { 0x454F46 }), writeData: 100),
            };
            Randomizer.CreateIPS(operations);
        }
    }
}

File: RewardRandomizer.Tests/IPSTests.fs
namespace RewardRandomizer.Tests

open System
open System.Diagnostics
open System.IO
open Microsoft.VisualStudio.TestTools.UnitTesting
open RewardRandomizer.Randomizer
open RewardRandomizer

[<TestClass>]
type IPSTests() =
    [<TestMethod>]
    member _.TestGeneration() =
        let random = new Random()
        let data1 = Array.zeroCreate<byte> 0x100
        random.NextBytes data1
        let operations = [
            { Offsets = set [0x30; 0x50; 0x51]; WriteData = 7uy }
            { Offsets = set [0x9a]; WriteData = 100uy }
        ]
        if not (File.Exists "Lunar IPS.exe") then
            Assert.Inconclusive()
        File.WriteAllBytes("original_file", data1)
        File.WriteAllBytes("patch_file", Randomizer.CreateIPS operations)
        Process.Start("Lunar IPS.exe", "-ApplyIPS patch_file original_file").WaitForExit();
        let expected = File.ReadAllBytes "original_file"
        File.Delete "original_file"
        File.Delete "patch_file"
        let actual = Randomizer.ApplyOperations data1 operations
        Assert.AreEqual(List.ofArray expected, List.ofArray actual)

    [<TestMethod>]
    [<ExpectedException(typeof<Exception>)>]
    member _.TestProhibitAmbiguity() =
        let random = new Random()
        let data1 = Array.zeroCreate<byte> 0x500000
        random.NextBytes data1
        let operations = [{ Offsets = set [0x454F46]; WriteData = 100uy }]
        Randomizer.CreateIPS operations
